src/net/socket.rs

```diff
@@ -0,0 +1,222 @@
+use crate::config::NetworkConfig;
+use crate::error::{NetworkError, NetworkErrorKind, NetworkResult};
+use crate::net::link_conditioner::LinkConditioner;
+use crate::net::{connection::ActiveConnections, events::SocketEvent};
+use crate::packet::Packet;
+use log::error;
+use mio::{Evented, Events, Poll, PollOpt, Ready, Token};
+use std::io;
+use std::mem;
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::sync::{mpsc, Arc};
+
+const SOCKET: Token = Token(0);
+
+/// A reliable UDP socket implementation with configurable reliability and ordering guarantees.
+pub struct LaminarSocket {
+    socket: mio::net::UdpSocket,
+    config: Arc<NetworkConfig>,
```

Comment:
Yes, as long it is used at more places it is better to do so. We did this because we were making multiple copies of it.

Reply:
right what im saying is we should analyze to see if we even need this as it may just be uneeded overhead.

Reply:
I mostly wanted to minimize the changes so I left this. I did remove it in my refactor, however. I can remove it if we want.

Reply:
👍 I suspect we can remove almost all these arcs since we sit within one thread now instead of multiple. So I think as a part of this change we should try and remove our dependence on `std::sync` as much as possible.

Reply:
I'll do that tonight. 👍